dashboard/lib/services/script_seed.rb

```diff
@@ -0,0 +1,320 @@
+  #
+  # @param [String | File] file_or_path - Can be String representing a path, relative or absolute, to the file
+  #   to read from, or it can be a File object to read from.
+  # @return [Script] the Script created/updated from seeding
+  def self.seed_from_json_file(file_or_path)
+    seed_from_json(File.read(file_or_path))
+  end
+
+  # Creates / updates the objects in the database described by the input JSON.
+  #
+  # This method is responsible for Script objects and everything "under" them logically in the curriculum data
+  # hierarchy. Currently (9/23/2020), this looks like:
+  #
+  # Script (has_many)-> LessonGroup (has_many)-> Lesson (has_many)-> ScriptLevel.
+  #
+  # This method is not responsible for creating/updating Levels, so it depends on level seeding running before it.
+  # However, it is responsible for creating/updating the associations between ScriptLevels and Levels, which control
+  # which Levels belong to which Lessons. Since ScriptLevel/Levels is a many-to-many relationship, controlled by a
+  # separate join table, we also treat the join table objects as belonging to the hierarchy, and represent it with
+  # the LevelsScriptLevels model.
+  #
+  # An outline of the approach:
+  #
+  # Go through the hierarchy top-to-bottom order, starting with Script, and do the following for each model:
+  # 1. For each object for the current model, look up and fill in the ids for their associated objects, based on their
+  # seeding keys. The rest of the attributes are used as-is from the input data.
+  # 2. Bulk import all of those objects using ActiveRecord-Import
+  # 3. After import, query for all objects of that model now associated with the Script. Find all of these objects which
+  # were not present in the input data, based on their seeding keys. Bulk-destroy those objects.
+  #
+  # Some important implementation details:
+  #
+  # - One challenge is that different models have different ways of uniquely identifying themselves. To abstract these
+  # differences, each model implements a seeding_key method which takes in a SeedContext, and returns a hash of all
+  # the information which composes its "logical unique identifier".
+  #
+  # - We try to avoid doing database queries within loops, within reason. Ideally, the total number of queries made
+  # would only scale with the # of models, not the number of objects. Exceptions can be made if avoiding them is
+  # too complex or not performant, and making the repeated queries achieves decent performance empirically.
+  #
+  # - One way to avoid repeated queries when looking up associations is to load all objects of the associated model
+  # for the Script into the SeedContext, and do lookups against that already-loaded data.
+  #
+  # - If a seeding_key has a corresponding unique index, we can skip looking up objects to update, and just
+  # rely on the "on duplicate key update" feature instead.
```

Comment:
are script levels the only model here which can't rely on "on duplicate key update" for this reason?

Reply:
Yep. And hopefully we'll fix that too with the seeding_key GUID.